lib/src/git.rs

```diff
@@ -1330,65 +1335,191 @@ fn push_refs(
+        if remaining_remote_refs.is_empty() {
+            Ok(())
+        } else {
+            Err(GitPushError::RefUpdateRejected(
+                remaining_remote_refs
+                    .iter()
+                    .sorted()
+                    .map(|name| name.to_string())
+                    .collect(),
+            ))
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum PushAllowReason {
+    NormalMatch,
+    ExceptionalFastforward,
+    UnexpectedNoop,
+}
+
+fn allow_push(
+    index: &dyn Index,
+    actual_remote_location: Option<&CommitId>,
+    expected_remote_location: Option<&CommitId>,
+    local_location: Option<&CommitId>,
+) -> Result<PushAllowReason, ()> {
+    if actual_remote_location == expected_remote_location {
+        return Ok(PushAllowReason::NormalMatch);
+    }
+
+    // If the remote ref is in an unexpected location, we still allow some
+    // pushes, based on whether `jj git fetch` would result in a conflicted
+    // ref.
+    //
+    // For `merge_ref_targets` to work, `actual_remote_location` must be a commit
+    // that we locally know about. This does not lose any generality since for
+    // `merge_ref_targets` to resolve to `local_target`, it is
+    // necessary (but not sufficient) for the push to be a fast-forward or a noop:
+    // local_target is either a descendant of actual_remote_location or equal to it.
+    // TODO(ilyagr): Use if `Some(id) = ... && ...` when that stabilizes.
+    if actual_remote_location
+        .map(|id| !index.has_id(id))
+        .unwrap_or(false)
+    {
+        return Err(());
+    }
+    let remote_target = RefTarget::resolved(actual_remote_location.cloned());
+    let base_target = RefTarget::resolved(expected_remote_location.cloned());
+    let local_target = RefTarget::resolved(local_location.cloned());
+    if refs::merge_ref_targets(index, &remote_target, &base_target, &local_target) != local_target {
+        return Err(());
+    }
+
+    // Fetch would not change the local branch, so the push is OK in spite of
+    // the discrepancy with the expected location. We return some debug info and
+    // verify some invariants before OKing the push.
+
+    if remote_target == local_target {
+        // This is the situation of what we call "A - B + A = A"
+        // conflicts, see also test_refs.rs and
+        // https://github.com/martinvonz/jj/blob/c9b44f382824301e6c0fdd6f4cbc52bb00c50995/lib/src/merge.rs#L92.
+        Ok(PushAllowReason::UnexpectedNoop)
+    } else {
+        // The assertions follow from our ref merge rules.
+        debug_assert!(index.is_ancestor(
+            actual_remote_location.as_ref().unwrap(),
+            local_location.as_ref().unwrap()
+        ));
+        debug_assert!(
+            expected_remote_location.is_none()
+                || index.is_ancestor(
+                    expected_remote_location.unwrap(),
+                    actual_remote_location.as_ref().unwrap()
+                )
+        );
```

Comment:
nit: I don't think these debug assertions help. Maybe this can be reorganized as follows in order to make it clear that `} else {` means fast-forward?

```rust
if merge_ref_targets(...) == local_target { // means it is fast forward
    if remote_target == local_target {  // special case of fast-forward
        UnexpectedNoop
    } else {
        ExceptionalFastForward
    }
} else {
    Err  // not fast-forward
}
```

Reply:
I cleaned it up a bit more. I still feel conflicted about removing the "expected location is the ancestor of both other locations" assertion/explanation/verification.

Reply:
It was mainly because I found the logic confusing. but I'll trust your judgement that it didn't help.